launcher/src/main/java/org/apache/spark/launcher/ChildProcAppHandle.java

```diff
@@ -48,14 +48,16 @@ public synchronized void disconnect() {
 
   @Override
   public synchronized void kill() {
-    disconnect();
-    if (childProc != null) {
-      if (childProc.isAlive()) {
-        childProc.destroyForcibly();
+    if (!isDisposed()) {
+      setState(State.KILLED);
```

Comment:
Why put `setState` in the front? It should be OK but in previous code some of the `setState` is called in the end.

Reply:
I have the same question, but should be OK as we always synchronize when touching the state.

Reply:
I changed this because the old disconnect code, at least, might change the handle's state. It's easier to put this call first and not have to reason about whether that will happen.